File: LeetCode/234.Palindrome_Linked_List/Mario/234.Palindrome_Linked_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        // 1->2->3->4
        // get half end
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode slow = dummy, fast = dummy;
        while (fast != null) {
            slow = slow.next;

            fast = fast.next;
            if (fast != null) {
                fast = fast.next;
            }
        }

        // reverse half end linkedlist
        ListNode other = null;
        while (slow != null) {
            ListNode tmp = slow.next;
            slow.next = other;
            other = slow;
            slow = tmp;
        }

        // compare it
        ListNode cur = dummy.next;
        while (other != null && cur != null) {
            if (other.val != cur.val) {
                return false;
            } else {
                other = other.next;
                cur = cur.next;
            }
        }

        return true;
    }
}

File: LeetCode/234.Palindrome_Linked_List/Mario/234.Palindrome_Linked_List.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func isPalindrome(_ head: ListNode?) -> Bool {
        guard let head = head, head.next != nil else { return true }

        // fast, slow pointers
        var fast: ListNode? = head, slow: ListNode? = head
        while fast != nil, fast?.next != nil {
            slow = slow?.next
            fast = fast?.next?.next
        }
        // if length of this List is odd
        if fast != nil {
            slow = slow?.next
        }

        var right = reverse(slow)
        var left: ListNode? = head

        while right != nil {
            if right?.val != left?.val {
                return false
            }
            right = right?.next
            left = left?.next
        }

        return true
    }

    private func reverse(_ node: ListNode?) -> ListNode? {
        var tail: ListNode? = nil
        var cur = node

        while let c = cur {
            let next = c.next
            c.next = tail
            tail = c
            cur = next
        }

        return tail
    }
}
